corehq/apps/users/models.py

```diff
@@ -2125,20 +2189,33 @@ def update_device_id_last_used(self, device_id, when=None):
         """
         when = when or datetime.utcnow()
 
-        for user_device_id_last_used in self.devices:
-            if user_device_id_last_used.device_id == device_id:
-                if when.date() > user_device_id_last_used.last_used.date():
-                    user_device_id_last_used.last_used = when
-                    return True
-                else:
-                    return False
+        device = self.get_device(device_id)
+        if device:
+            if when.date() > device.last_used.date():
+                device.last_used = when
+                device.update_meta(commcare_version, device_app_meta)
+                return True
+            else:
+                return False
         else:
-            self.devices.append(DeviceIdLastUsed(
-                device_id=device_id,
-                last_used=when
-            ))
+            device = DeviceIdLastUsed(device_id=device_id, last_used=when)
+            device.update_meta(commcare_version, device_app_meta)
+            self.devices.append(device)
             return True
 
+    def get_last_used_device(self):
+        if not self.devices:
+            return None
+
+        return sorted(self.devices, key=lambda dev: dev.last_used)[-1]
+
+    def get_device(self, device_id):
+        matches = [
```

Comment:
This could return early if written like (I don't htink python has a `first` method):

```
for device in devices:
    if device.device_id == device_id:
        return device

return None
```